core/common/src/LocalDateRange.kt

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2019-2022 JetBrains s.r.o.
+ * Use of this source code is governed by the Apache 2.0 License that can be found in the LICENSE.txt file.
+ */
+
+package kotlinx.datetime
+
+import kotlin.random.Random
+import kotlin.random.nextInt
+
+internal class LocalDateProgressionIterator(private val iterator: IntIterator) : Iterator<LocalDate> {
+    override fun hasNext(): Boolean = iterator.hasNext()
+    override fun next(): LocalDate = LocalDate.fromEpochDays(iterator.next())
+}
+
+public open class LocalDateProgression
+internal constructor(internal val intProgression: IntProgression) : Iterable<LocalDate> {
+
+    internal constructor(
+        start: LocalDate,
+        endInclusive: LocalDate,
+        step: Int
+    ) : this(IntProgression.fromClosedRange(start.toEpochDays(), endInclusive.toEpochDays(), step))
+
+    public val first: LocalDate = LocalDate.fromEpochDays(intProgression.first)
+    public val last: LocalDate = LocalDate.fromEpochDays(intProgression.last)
+
+    override fun iterator(): Iterator<LocalDate> = LocalDateProgressionIterator(intProgression.iterator())
+
+    public open fun isEmpty(): Boolean = intProgression.isEmpty()
+
+    override fun toString(): String = if (intProgression.step > 0) "$first..$last step ${intProgression.step}" else "$first downTo $last step ${intProgression.step}"
```

Comment:
If you call `toString` on other range implementations, the result can be copy-pasted into the code to obtain the same progression. This one can not, and I think it would be nice to preserve that property.

Reply:
This is the one place where `step` is semi-exposed publicly. Since step isn't really settled, I didn't want to commit too much in any direction.

Being able to copy-paste into code is a nice consequence of the code representation of primitives matching their string representation, but they diverge for more complex objects. Additionally, there is no infix step function for `LocalDateProgression` yet. I think we would lose a lot in readability if we were to make it copy-pasteable:

`LocalDate.parse("2000-01-01")..LocalDate.parse("2000-01-10").step(2, DateTimeUnit.DAY)` is a pretty ugly string representation. We could improve it slightly with:

`LocalDate(2000, 1, 1)..LocalDate(2000, 1, 10).step(2, DateTimeUnit.DAY)`, but it still has poor readability, and at that point we're abandoning the existing string representation of `LocalDate` entirely.

Reply:
Ok, makes sense, thanks! Could you then include the information that the step is measured in days?